src/poly/dcrt/matrix/dcrt_poly.rs

```diff
@@ -202,72 +208,83 @@ impl PolyMatrix for DCRTPolyMatrix {
-    //                     );
-    //                     let mut path = dir_path.as_ref().to_path_buf();
-    //                     path.push(format!(
-    //                         "{}_{}.{}_{}.{}.matrix",
-    //                         block_size,
-    //                         cur_block_row_idx,
-    //                         next_block_row_idx,
-    //                         cur_block_col_idx,
-    //                         next_block_col_idx
-    //                     ));
-    //                     let entries_bytes: Vec<Vec<Vec<u8>>> = entries
-    //                         .iter()
-    //                         .map(|row| row.iter().map(|poly|
-    // poly.to_compact_bytes()).collect_vec())                         .collect_vec();
-    //                     serde_json::to_writer(
-    //                         std::fs::File::create(&path).unwrap(),
-    //                         &entries_bytes,
-    //                     )
-    //                     .unwrap_or_else(|_| panic!("Failed to write matrix file {:?}", path));
-    //                 },
-    //             );
-    //         },
-    //     );
-    // }
+    fn read_from_files<P: AsRef<Path> + Send + Sync>(
+        params: &<Self::P as Poly>::Params,
+        nrow: usize,
+        ncol: usize,
+        dir_path: P,
+        id: &str,
+    ) -> Self {
+        let block_size = block_size();
+        let mut matrix = Self::new_empty(params, nrow, ncol);
+
+        let (row_offsets, col_offsets) = block_offsets(0..nrow, 0..ncol);
+
+        for (row_start, row_end) in row_offsets.iter().tuple_windows() {
+            for (col_start, col_end) in col_offsets.iter().tuple_windows() {
+                let f = |row_range: Range<usize>, col_range: Range<usize>| -> Vec<Vec<DCRTPoly>> {
+                    let mut path = dir_path.as_ref().to_path_buf();
+                    path.push(format!(
+                        "{}_{}_{}.{}_{}.{}.matrix",
+                        id, block_size, row_start, row_end, col_start, col_end
+                    ));
+                    let bytes = std::fs::read(&path)
+                        .unwrap_or_else(|_| panic!("Failed to read matrix file {:?}", path));
+                    let entries_bytes: Vec<Vec<Vec<u8>>> = serde_json::from_slice(&bytes).unwrap();
+
+                    parallel_iter!(0..row_range.len())
+                        .map(|i| {
+                            parallel_iter!(0..col_range.len())
+                                .map(|j| {
+                                    let entry_bytes = &entries_bytes[i][j];
+                                    DCRTPoly::from_compact_bytes(params, entry_bytes)
+                                })
+                                .collect::<Vec<_>>()
+                        })
+                        .collect::<Vec<_>>()
+                };
+
+                matrix.replace_entries(*row_start..*row_end, *col_start..*col_end, f);
+            }
+        }
```

Comment:
Why does this code call `replace_entries` for each chunk of (row_start..row_end, col_start..col_end)?
I think the same process is done within the  `replace_entries` function.